File: pachi/pachi-rust/src/ball_sink.rs
use std::io::sink;

use godot::{
    classes::{Area2D, IMarker2D, Marker2D},
    prelude::*,
};

use crate::ball::Ball;

#[derive(GodotClass)]
#[class(init, base=Marker2D)]
pub struct BallSink {
    #[export]
    area: Option<Gd<Area2D>>,

    base: Base<Marker2D>,
}

#[godot_api]
impl IMarker2D for BallSink {
    fn ready(&mut self) {
        assert!(self.area.is_some());
    }

    fn physics_process(&mut self, _delta: f64) {
        for overlapper in self
            .area
            .as_ref()
            .unwrap()
            .get_overlapping_bodies()
            .iter_shared()
        {
            let mut ball = match overlapper.try_cast::<Ball>() {
                Ok(ball) => ball,
                _ => continue,
            };

            // TODO: read the correct fields, don't just hardcode
            let sink_radius = 16.;
            let ball_radius = 10.;
            let distance =
                (self.base().get_global_position() - ball.get_global_position()).length();
            if distance <= sink_radius - ball_radius {
                // TODO: signalling
                self.signals().ball_sunk().emit(&ball);
                ball.queue_free();
            }
        }
    }
}

#[godot_api]
impl BallSink {
    #[signal]
    fn ball_sunk(ball: Gd<Ball>);
}

File: pachi.04/content/ball_sink/BallSink.cs
using Godot;

public partial class BallSink : Marker2D
{
    [Signal]
    public delegate void BallSunkEventHandler(Ball ball);

    [Export]
    public Area2D Area { get; set; }

    public override void _Ready()
    {
        if (Area == null)
        {
            GD.PushError("BallSink: Area is not assigned!");
        }
    }

    public override void _PhysicsProcess(double delta)
    {
        if (Area == null) return;

        foreach (Node body in Area.GetOverlappingBodies())
        {
            if (body is Ball ball)
            {
                // TODO: read the correct fields, don't just hardcode
                float sinkRadius = 16.0f;
                float ballRadius = 10.0f;
                float distance = (GlobalPosition - ball.GlobalPosition).Length();

                if (distance <= sinkRadius - ballRadius)
                {
                    EmitSignal(SignalName.BallSunk, ball);
                    ball.QueueFree();
                }
            }
        }
    }
}
